extensions/src/org/pentaho/platform/web/http/api/resources/services/UserRoleListService.java

```diff
@@ -74,7 +74,22 @@ public RoleListWrapper getRoles() {
 
   public RoleListWrapper getAllRoles() {
     List<String> roles = getUserRoleListService().getAllRoles();
-    roles.addAll( getExtraRoles() );
+    List<String> extraRoles = getExtraRoles();
+    if ( extraRoles != null ) {
+      if ( systemRoles != null ) {
+        extraRoles.addAll( systemRoles );
+      }
+    }
+    else {
+      extraRoles = systemRoles;
+    }
+    if ( extraRoles != null ) {
+      for ( String ex : extraRoles ) {
```

Comment:
This would be way faster using a HashSet. Especially if the number of roles is large (like for instance, a large LDAP installation). I strongly suggest re-working this to not use contains and a list, and instead, use a HashSet.